### DISCUSSION TITLE
Xamarin.iOS RabbitMQ_Client_Impl_SocketFrameHandler_Close SocketFrameHandler.cs, line 196

### DISCUSSION THREAD

**Discussion by VBoydev** [rabbitmq/rabbitmq-dotnet-client#1102] [General]
Hi,

We do get the fallowing exception which started happening after we started using SSL, the exception is when we are trying to initialise the connection and the app is crashing. It is a an exception trown in the RabitMq library and the try catch is not helping. The question maybe is any idea how we can catch this exception so the app is not crashing or anything else that mihght fix the issue. 

Here is the exception:

Stacktrace

RabbitMQ_Client_Impl_SocketFrameHandler_Close (SocketFrameHandler.cs:196);Vapp.iOS
RabbitMQ_Client_Impl_SocketFrameHandler__ctor_RabbitMQ_Client_AmqpTcpEndpoint_System_Func_2_System_Net_Sockets_AddressFamily_RabbitMQ_Client_ITcpClient_System_TimeSpan_System_TimeSpan_System_TimeSpan (SocketFrameHandler.cs:126);Vapp.iOS
RabbitMQ_Client_Framing_Impl_IProtocolExtensions_CreateFrameHandler_RabbitMQ_Client_IProtocol_RabbitMQ_Client_AmqpTcpEndpoint_System_Func_2_System_Net_Sockets_AddressFamily_RabbitMQ_Client_ITcpClient_System_TimeSpan_System_TimeSpan_System_TimeSpan (IProtocolExtensions.cs:50);Vapp.iOS
RabbitMQ_Client_ConnectionFactory_CreateFrameHandler_RabbitMQ_Client_AmqpTcpEndpoint (ConnectionFactory.cs:513);Vapp.iOS
RabbitMQ_Client_EndpointResolverExtensions_SelectOne_T_REF_RabbitMQ_Client_IEndpointResolver_System_Func_2_RabbitMQ_Client_AmqpTcpEndpoint_T_REF (IEndpointResolverExtensions.cs:47);Vapp.iOS
RabbitMQ_Client_Framing_Impl_AutorecoveringConnection_Init_RabbitMQ_Client_IEndpointResolver (AutorecoveringConnection.cs:635);Vapp.iOS
RabbitMQ_Client_ConnectionFactory_CreateConnection_RabbitMQ_Client_IEndpointResolver_string (ConnectionFactory.cs:494);Vapp.iOS
RabbitMQ_Client_ConnectionFactory_CreateConnection_string (ConnectionFactory.cs:380);Vapp.iOS
RabbitMQ_Client_ConnectionFactory_CreateConnection (ConnectionFactory.cs:361);Vapp.iOS
Vapp_Services_Network_QueueApi_WebMessageClient__InitializeDownloadMessageConnectionAsyncd__21_MoveNext (:1);Vapp.iOS
System_Runtime_CompilerServices_AsyncTaskMethodBuilder_Start_Vapp_Services_Network_QueueApi_WebMessageClient__InitializeDownloadMessageConnectionAsyncd__21_Vapp_Services_Network_QueueApi_WebMessageClient__InitializeDownloadMessageConnectionAsyncd__21_ (AsyncMethodBuilder.cs:316);Vapp.iOS
Vapp_Services_Network_QueueApi_WebMessageClient_InitializeDownloadMessageConnectionAsync_bool (:1);Vapp.iOS
ut_end (:1);Vapp.iOS
System_Runtime_CompilerServices_AsyncTaskMethodBuilder_Start_Vapp_App___OnResumeb__30_0d_Vapp_App___OnResumeb__30_0d_ (AsyncMethodBuilder.cs:316);Vapp.iOS
Vapp_App__OnResumeb__30_0 (:1);Vapp.iOS
System_Threading_Tasks_Task_1_TResult_REF_InnerInvoke + 84;Vapp.iOS
System_Threading_Tasks_Task_Execute + 32;Vapp.iOS
System_Threading_Tasks_Task_ExecutionContextCallback_object + 84;Vapp.iOS
System_Threading_ExecutionContext_RunInternal_System_Threading_ExecutionContext_System_Threading_ContextCallback_object_bool + 424;Vapp.iOS
System_Threading_ExecutionContext_Run_System_Threading_ExecutionContext_System_Threading_ContextCallback_object_bool + 40;Vapp.iOS
System_Threading_Tasks_Task_ExecuteWithThreadLocal_System_Threading_Tasks_Task_ + 292;Vapp.iOS
System_Threading_Tasks_Task_ExecuteEntry_bool + 268;Vapp.iOS
System_Threading_Tasks_Task_System_Threading_IThreadPoolWorkItem_ExecuteWorkItem + 20;Vapp.iOS
System_Threading_ThreadPoolWorkQueue_Dispatch + 484;Vapp.iOS
ObjCRuntime_Runtime_ThreadPoolDispatcher_System_Func_1_bool (Runtime.cs:330);Vapp.iOS
System_Threading__ThreadPoolWaitCallback_PerformWaitCallback + 132;Vapp.iOS
wrapper_runtime_invoke_object_runtime_invoke_dynamic_intptr_intptr_intptr_intptr + 268;Vapp.iOS
mono_jit_runtime_invoke (mini-runtime.c:3190);Vapp.iOS
mono_runtime_try_invoke (object.c:3052);Vapp.iOS
worker_callback (threadpool.c:292);Vapp.iOS
worker_thread (threadpool-worker-default.c:502);Vapp.iOS
start_wrapper_internal (threads.c:1233);Vapp.iOS
start_wrapper (threads.c:1308);Vapp.iOS

Reason

SIGSEGV
![Screenshot 2021-11-08 at 12 01 43](https://user-images.githubusercontent.com/27084190/140722387-a185e606-c3c0-40b3-b986-ff0a4a1809f6.png)

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
This client does not generally target platforms beyond .NET Core and .NET Framework. In any case, it has no native code and cannot be responsible for a `SIGSEGV` (a segmentation fault).

**Reply by VBoydev** [reply]
Xamarin is part of the .Net Framework and the exception is triggered because of the error in SocketFrameHandler.cs and thus it is not catched from our try catch

**Comment by lukebakken** [comment]
> We do get the fallowing exception which started happening after we started using SSL

Can you confirm that everything works as intended _without_ using TLS? (SSL)

**Reply by VBoydev** [reply]
Yes, I can confirm that, as this was the case before the switch to TLS(SSL).

**Reply by VBoydev** [reply]
it is starting here: Line 122 of SocketFrameHandler -> netstream = SslHelper.TcpUpgrade(netstream, endpoint.Ssl); and then the Close() is called on line 126 
![Screenshot 2021-11-09 at 8 54 00](https://user-images.githubusercontent.com/27084190/140876438-950271bd-e7c4-40b0-b249-5f05a15089d1.png)

The Close() method is called which contains line 196: -> _writerTask.GetAwaiter().GetResult();
![Screenshot 2021-11-09 at 8 54 24](https://user-images.githubusercontent.com/27084190/140876827-7eeef9aa-bb10-49fb-a1a7-4d38bb1f520c.png)

What we think is that: writerTask on line 196 is null at this moment (it gets initialized later in the constructor in the happy case) we get a NullReferenceException. There is a try catch around this line but as we know NullReferenceExceptions cause SIGSEV app crashes in  release mode of Xamarin.iOS.
![Screenshot 2021-11-09 at 9 06 12](https://user-images.githubusercontent.com/27084190/140877765-1b3531e2-7440-4325-8586-a992b251b823.png)

_writerTask gets initialized on line 135 after we throw the exception and call Close();

**Comment by lukebakken** [comment]
@VBoydev - you have a good understanding of the issue. Please test changes to fix this and open a pull request. We do not have access to a device running iOS to test this on.

**Reply by VBoydev** [reply]
Fyi we are testing the fix and I will do PR as soon as we are sure that it is working.

**Comment by VBoydev** [comment]
Hi we have an issue when we are doing the tests in release mode we can't establish a connection, we do not have this issue when testing our app in debug mode. Also when building the nuget we use MassPublish -> Release, is this the correct way to build the nuget?

**Reply by VBoydev** [reply]
@michaelklishin @lukebakken

**Reply by lukebakken** [reply]
There is no need to ping us directly. We provide free assistance when time allows. Since it sounds like you're having a different issue than the original one I suggest starting a new discusson. Please provide more information. A sentence describing your issue is not sufficient. Logs, exceptions would be helpful. I do not have any way to test the .NET client using Xamarin so I am 100% dependent on information from you.
